src/ImmutableProperties.php

```diff
@@ -0,0 +1,120 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Codelicia\Immutable;
+
+use ReflectionObject;
+use ReflectionProperty;
+use ReflectionType;
+use RuntimeException;
+use function array_key_exists;
+use function class_implements;
+use function get_class;
+use function gettype;
+use function in_array;
+use function interface_exists;
+use function is_object;
+use function spl_object_hash;
+use function sprintf;
+
+trait ImmutableProperties
+{
+    /** @return string[][]|ReflectionProperty[][] */
+    protected function &reflectionProperty(): array
+    {
+        static $reflectionProperty = [];
+
+        if (! array_key_exists(spl_object_hash($this), $reflectionProperty)) {
+            $reflectionProperty[spl_object_hash($this)] = [];
+        }
+
+        return $reflectionProperty[spl_object_hash($this)];
+    }
+
+    public function __construct()
+    {
+        $this->init();
+    }
+
+    public function init(): void
+    {
+        $reflection = new ReflectionObject($this);
+        $store = &$this->reflectionProperty();
+        $defaultProperties = $reflection->getDefaultProperties();
+
+        foreach ($reflection->getProperties(ReflectionProperty::IS_PUBLIC) as $reflectionProperty) {
+            $propertyName = $reflectionProperty->name;
+            /** @var \ReflectionType $type */
+            $type = $reflectionProperty->getType();
+
+            $store[$propertyName] = [
+                'reflection' => $reflectionProperty,
+                'value' => $defaultProperties[$propertyName] ?: null,
+                'isInitialized' => isset($defaultProperties[$propertyName]) ? true : $reflectionProperty->isInitialized($this),
+            ];
+
+            if (isset($defaultProperties[$propertyName]) || $type->allowsNull()) {
+                unset($this->{$propertyName});
+            }
+        }
+    }
+
+    public function __set(string $prop, $value)
+    {
+        $ref = &$this->reflectionProperty();
```

Comment:
Wouldn't it be simpler to check if `array_key_exists($nameOfProperty, (array) $this)` to see if the property is already initialized? That would avoid a lot of copying references.

Reply:
Actually can't as typed properties still exists even when `unset()` is being used on it.

```php
<?php

declare(strict_types=1);

class User { public string $name; }

$user = new User;
var_dump($user);

//object(User)#1 (0) {
//["name"]=>
//  uninitialized(string)
//}

$user->name = 'foo';
var_dump($user);

//object(User)#1 (1) {
//["name"]=>
//  string(3) "foo"
//}

unset($user->name);
var_dump($user);

//object(User)#1 (0) {
//["name"]=>
//  uninitialized(string)
//}
```